File: RamanControl.py
import numpy as np
from types import MethodType, FunctionType
from Wrapper import *


class RamanControl:

    def __init__(self, **kwargs):
        """
        __init__ function call to initialize variables from the
        parameters for the class instance provided in __main__ and
        add new variables for use in other functions in this class.
        """

        for name, value in kwargs.items():
            if isinstance(value, FunctionType):
                setattr(self, name, MethodType(value, self))
            else:
                setattr(self, name, value)

        self.dyn_rho_A = np.empty((int(self.nDIM * (self.nDIM + 3) / 2), self.timeDIM), dtype=np.complex)
        self.dyn_rho_B = np.empty_like(self.dyn_rho_A, dtype=np.complex)

        self.rho_A = np.empty((self.nDIM, self.nDIM), dtype=np.complex)
        self.rho_B = np.empty_like(self.rho_A, dtype=np.complex)

        self.time = np.linspace(-self.timeAMP, self.timeAMP, self.timeDIM, endpoint=False)
        self.time_spectra = np.linspace(-self.timeAMP_spectra, self.timeAMP_spectra, self.timeDIM_spectra, endpoint=False)
        self.field = np.empty(self.timeDIM)
        self.field_spectra = np.empty(self.timeDIM_spectra, dtype=np.complex)
        self.freq_spectra = np.linspace(self.freq_min, self.freq_max, self.freqDIM)
        self.spectra_A = np.empty(self.freqDIM, dtype=np.complex)
        self.spectra_B = np.empty(self.freqDIM, dtype=np.complex)

    def create_molecule(self, MolA, MolB):
        """
        Initializes two instance of the Structure "Molecule"
        :param MolA: Molecule A of type Structure(Molecule)
        :param MolB: Molecule B of type Structure(Molecule)
        :return: None
        """

        MolA.energies = self.energies_A.ctypes.data_as(POINTER(c_double))
        MolB.energies = self.energies_B.ctypes.data_as(POINTER(c_double))
        MolA.gamma_decay = self.gamma_decay_A.ctypes.data_as(POINTER(c_double))
        MolB.gamma_decay = self.gamma_decay_B.ctypes.data_as(POINTER(c_double))
        MolA.gamma_pcd = self.gamma_pcd_A.ctypes.data_as(POINTER(c_double))
        MolB.gamma_pcd = self.gamma_pcd_B.ctypes.data_as(POINTER(c_double))
        MolA.rho = self.rho_A.ctypes.data_as(POINTER(c_complex))
        MolB.rho = self.rho_B.ctypes.data_as(POINTER(c_complex))
        MolA.dyn_rho = self.dyn_rho_A.ctypes.data_as(POINTER(c_complex))
        MolB.dyn_rho = self.dyn_rho_B.ctypes.data_as(POINTER(c_complex))
        MolA.spectra = self.spectra_A.ctypes.data_as(POINTER(c_complex))
        MolB.spectra = self.spectra_B.ctypes.data_as(POINTER(c_complex))

        return

    def create_parameters(self, func_params):
        """
        Initializes an instance of the Structure "Parameters"
        :param func_params: Parameter Structure instance to be updated
        :return: None
        """

        func_params.time = self.time.ctypes.data_as(POINTER(c_double))
        func_params.rho_0 = self.rho_0.ctypes.data_as(POINTER(c_complex))
        func_params.mu = self.mu.ctypes.data_as(POINTER(c_complex))
        func_params.A_R = self.A_R
        func_params.width_R = self.timeAMP / self.width_R
        func_params.t0_R = self.timeAMP * self.t0_R
        func_params.A_EE = self.A_EE
        func_params.width_EE = self.timeAMP / self.width_EE
        func_params.t0_EE = self.timeAMP * self.t0_EE
        func_params.w_R = self.w_R
        func_params.w_v = self.w_v
        func_params.w_EE = self.w_EE
        func_params.nDIM = self.nDIM
        func_params.timeDIM = self.timeDIM
        func_params.field_out = self.field.ctypes.data_as(POINTER(c_double))

        return

    def create_spectra_parameters(self, spec_params):
        """
        Initializes an instance of the Structure "Parameters"
        :param func_params: Parameter Structure instance to be updated
        :return: None
        """

        spec_params.field_spectra = self.field_spectra.ctypes.data_as(POINTER(c_complex))
        spec_params.freq_spectra = self.freq_spectra.ctypes.data_as(POINTER(c_double))
        spec_params.freqDIM = self.freqDIM
        spec_params.w_R = self.w_R
        spec_params.A_S = self.A_S
        spec_params.width_S = self.timeAMP_spectra / self.width_S
        spec_params.time_spectra = self.time_spectra.ctypes.data_as(POINTER(c_double))
        spec_params.timeDIM_spectra = self.timeDIM_spectra
        spec_params.nDIM = self.nDIM

        return

    def call_raman_optimization(self):
        """

        :return:
        """
        MolA = Molecule()
        MolB = Molecule()
        func_params = Parameters()
        spec_params = SpectraParams()
        self.create_molecule(MolA, MolB)
        self.create_parameters(func_params)
        self.create_spectra_parameters(spec_params)
        RamanControlFunction(MolA, MolB, func_params, spec_params)


if __name__ == '__main__':

    import matplotlib.pyplot as plt

    np.set_printoptions(precision=4)

    energy_factor = 1. / 27.211385
    time_factor = .02418
    timeAMP = 20000.
                                                                                ##################################################################
                                                                                #                                                                #
    energies_A = np.array((0.000, 0.07439, 1.94655, 2.02094)) * energy_factor   #   Energies in a.u. for molecule A                              #
    energies_B = np.array((0.000, 0.09639, 1.92655, 2.00094)) * energy_factor   #   Energies in a.u. for molecule B                              #
                                                                                #                                                                #
    N = len(energies_A)                                                         #                                                                #
    rho_0 = np.zeros((N, N), dtype=np.complex)                                  #                                                                #
    rho_0[0, 0] = 1. + 0j                                                       #   Only ground state populated                                  #
    mu = 4.97738 * np.ones_like(rho_0)                                          #   Dipole moment in a.u. (same for all allowed transition)      #
    np.fill_diagonal(mu, 0j)                                                    #   \mu_{i, i} = 0.                                              #
                                                                                #                                                                #
    population_decay = 2.418884e-8                                              #   All population relaxation lifetimes equal 1 ns               #
    electronic_dephasing = 3. * 2.418884e-4                                     #   All electronic dephasing 50 fs                               #
    vibrational_dephasing = 1.422872e-5                                         #   All vibrational dephasing 1 ps                               #
                                                                                #                                                                #
    gamma_decay = np.ones((N, N)) * population_decay                            #                                                                #
    np.fill_diagonal(gamma_decay, 0.0)                                          #   No decay for diagonal elements                               #
    gamma_decay = np.tril(gamma_decay)                                          #   Decay non-zero only from higher to lower energy levels       #
                                                                                #                                                                #
    gamma_pure_dephasing = np.ones_like(gamma_decay) * electronic_dephasing     #                                                                #
    np.fill_diagonal(gamma_pure_dephasing, 0.0)                                 #   No dephasing for population terms                            #
    gamma_pure_dephasing[1, 0] = vibrational_dephasing                          #                                                                #
    gamma_pure_dephasing[3, 2] = vibrational_dephasing                          ##################################################################
    gamma_pure_dephasing[0, 1] = vibrational_dephasing
    gamma_pure_dephasing[2, 3] = vibrational_dephasing
    gamma_pure_dephasing[2, 0] *= 1.
    gamma_pure_dephasing[0, 2] *= 1.
    gamma_pure_dephasing[3, 0] *= 2.
    gamma_pure_dephasing[0, 3] *= 2.
    gamma_pure_dephasing[1, 3] *= 1.
    gamma_pure_dephasing[3, 1] *= 1.

    lower_bounds = np.asarray([0.00020, timeAMP / 25.0, 0.9 * energies_A[1], 1.2 * energy_factor])
    upper_bounds = np.asarray([0.00070, timeAMP / 6.0, 1.1 * energies_A[1], 1.6 * energy_factor])

    guess = np.asarray([np.random.uniform(lower_bounds[i], upper_bounds[i]) for i in range(len(lower_bounds))])

    params = dict(
        energy_factor=energy_factor,
        time_factor=time_factor,
        timeDIM=40000,
        timeAMP=timeAMP,
        nDIM=N,

        A_R=0.000576595,
        width_R=5.017,
        t0_R=-0.35,

        A_EE=0.000366972,
        width_EE=19.32,
        t0_EE=0.55,

        w_R=1.4 * energy_factor,
        w_v=energies_A[1],
        w_EE=(energies_A[2] - energies_A[1]),
        rho_0=rho_0,

        lower_bounds=lower_bounds,
        upper_bounds=upper_bounds,
        guess=guess,

        energies_A=energies_A,
        energies_B=energies_B,
        gamma_decay_A=gamma_decay,
        gamma_decay_B=gamma_decay,
        gamma_pcd_A=gamma_pure_dephasing,
        gamma_pcd_B=gamma_pure_dephasing,
        mu=mu,

        freqDIM=100,
        freq_min=.05 * energy_factor,
        freq_max=.12 * energy_factor,
        A_S=0.00005,
        width_S=4.,
        timeDIM_spectra=40000,
        timeAMP_spectra=20000
    )

    gamma = 9. * gamma_pure_dephasing.copy()
    for i in range(4):
        for j in range(4):
            for k in range(4):
                if k > i:
                    gamma[i, j] += gamma_decay[k, i]
                if k > j:
                    gamma[i, j] += gamma_decay[k, j]
            gamma[i, j] *= 0.5

    print(gamma)
    N = 100
    omega = np.linspace(params['freq_min'], params['freq_max'], params['freqDIM'])
    spectraA = np.zeros(N)
    spectraB = np.zeros(N)

    for i in range(N):
        spectraA[i] *= 0.
        spectraB[i] *= 0.
        for j in range(1, 4):
            # spectraA[i] += energies_A[j] * np.abs(mu[j, 0]) ** 2 * gamma[j, 0] / (
            #         (energies_A[j] - omega[i]) ** 2 + gamma[j, 0] ** 2)
            # spectraB[i] += energies_B[j] * np.abs(mu[j, 0]) ** 2 * gamma[j, 0] / (
            #         (energies_B[j] - omega[i]) ** 2 + gamma[j, 0] ** 2)
            spectraA[i] += (energies_A[j] / (energies_A[j]**2 - (omega[i] - 1j*gamma[j, 0])**2)).imag * omega[i]
            spectraB[i] += (energies_B[j] / (energies_B[j]**2 - (omega[i] - 1j*gamma[j, 0])**2)).imag * omega[i]

    system = RamanControl(**params)
    system.call_raman_optimization()
    fig = plt.figure()
    plt.plot(system.time_spectra, system.field_spectra)

    fig1 = plt.figure()
    plt.plot(1239.84193 / system.freq_spectra * energy_factor, system.spectra_A.real / np.abs(system.spectra_A).max(), 'r')
    plt.plot(1239.84193 / system.freq_spectra * energy_factor, system.spectra_B.real / np.abs(system.spectra_A).max(), 'b')
    # plt.plot(1239.84193 / omega * energy_factor, -spectraA / np.abs(spectraA).max(), 'r--')
    # plt.plot(1239.84193 / omega * energy_factor, -spectraB / np.abs(spectraA).max(), 'b--')
    plt.show()

File: RamanControl.c
#include <complex.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <nlopt.h>

typedef double complex cmplx;

typedef struct parameters{
    double* time;
    cmplx* rho_0;
    cmplx* mu;

    double A_R;
    double width_R;
    double t0_R;

    double A_EE;
    double width_EE;
    double t0_EE;

    double w_R;
    double w_v;
    double w_EE;

    int nDIM;
    int timeDIM;

    double* field_out;
} parameters;

typedef struct spectra_params{
    cmplx* field_spectra;
    double* freq_spectra;
    int freqDIM;
    double w_R;
    double A_S;
    double width_S;
    double* time_spectra;
    int timeDIM_spectra;
    int nDIM;
} spectra_params;

typedef struct molecule{
    double* energies;
    double* gamma_decay;
    double* gamma_pcd;
    cmplx* rho;
    cmplx* dyn_rho;
    cmplx* spectra;
} molecule;

typedef struct mol_system{
    molecule* moleculeA;
    molecule* moleculeB;
    parameters* params;
} mol_system;


//====================================================================================================================//
//                                                                                                                    //
//                                        AUXILIARY FUNCTIONS FOR MATRIX OPERATIONS                                   //
//                                                                                                                    //
//====================================================================================================================//


void print_complex_mat(cmplx *A, int nDIM)
//----------------------------------------------------//
// 	          PRINTS A COMPLEX MATRIX                 //
//----------------------------------------------------//
{
	int i,j;
	for(i=0; i<nDIM; i++)
	{
		for(j=0; j<nDIM; j++)
		{
			printf("%3.3e + %3.3eJ  ", creal(A[i * nDIM + j]), cimag(A[i * nDIM + j]));
		}
	    printf("\n");
	}
	printf("\n\n");
}

void print_complex_vec(cmplx *A, int vecDIM)
//----------------------------------------------------//
// 	          PRINTS A COMPLEX MATRIX                 //
//----------------------------------------------------//
{
	int i;
	for(i=0; i<vecDIM; i++)
	{
		printf("%3.3e + %3.3eJ  ", creal(A[i]), cimag(A[i]));
	}
	printf("\n");
}

void print_double_mat(double *A, int nDIM)
//----------------------------------------------------//
// 	            PRINTS A REAL MATRIX                  //
//----------------------------------------------------//
{
	int i,j;
	for(i=0; i<nDIM; i++)
	{
		for(j=0; j<nDIM; j++)
		{
			printf("%3.3e  ", A[i * nDIM + j]);
		}
	    printf("\n");
	}
	printf("\n\n");
}

void print_double_vec(double *A, int vecDIM)
//----------------------------------------------------//
// 	          PRINTS A COMPLEX MATRIX                 //
//----------------------------------------------------//
{
	int i;
	for(i=0; i<vecDIM; i++)
	{
		printf("%3.3e  ", A[i]);
	}
	printf("\n");
}


void copy_mat(const cmplx *A, cmplx *B, int nDIM)
//----------------------------------------------------//
// 	        COPIES MATRIX A ----> MATRIX B            //
//----------------------------------------------------//
{
    int i, j = 0;
    for(i=0; i<nDIM; i++)
    {
        for(j=0; j<nDIM; j++)
        {
            B[i * nDIM + j] = A[i * nDIM + j];
        }
    }
}

void add_mat(const cmplx *A, cmplx *B, int nDIM)
//----------------------------------------------------//
// 	        ADDS A to B ----> MATRIX B = A + B        //
//----------------------------------------------------//
{
    int i, j = 0;
    for(i=0; i<nDIM; i++)
    {
        for(j=0; j<nDIM; j++)
        {
            B[i * nDIM + j] += A[i * nDIM + j];
        }
    }
}


void scale_mat(cmplx *A, double factor, int nDIM)
//----------------------------------------------------//
// 	     SCALES A BY factor ----> MATRIX B = A + B    //
//----------------------------------------------------//
{
    for(int i=0; i<nDIM; i++)
    {
        for(int j=0; j<nDIM; j++)
        {
            A[i * nDIM + j] *= factor;
        }
    }
}


double complex_abs(cmplx z)
//----------------------------------------------------//
// 	            RETURNS ABSOLUTE VALUE OF Z           //
//----------------------------------------------------//
{

    return sqrt((creal(z)*creal(z) + cimag(z)*cimag(z)));
}


cmplx complex_trace(cmplx *A, int nDIM)
//----------------------------------------------------//
// 	                 RETURNS TRACE[A]                 //
//----------------------------------------------------//
{
    cmplx trace = 0.0 + I * 0.0;
    for(int i=0; i<nDIM; i++)
    {
        trace += A[i*nDIM + i];
    }
    printf("Trace = %3.3e + %3.3eJ  \n", creal(trace), cimag(trace));

    return trace;
}


void multiply_complex_mat(cmplx *product, const cmplx *A, const cmplx *B, const int nDIM)
//----------------------------------------------------//
// 	             RETURNS A*B MATRIX PRODUCT           //
//----------------------------------------------------//
{
    for (int i=0; i<nDIM; i++)
    {
        for (int j=0; j<nDIM; j++)
        {
            for (int k=0; k<nDIM; k++)
            {
                product[i*nDIM + j] += A[i*nDIM + k]*B[k*nDIM + j];
            }
        }
    }
}


void commute_complex_mat(cmplx *commutator, const cmplx *A, const cmplx *B, const int nDIM)
//-----------------------------------------------------------------//
// 	          RETURNS commutator = [A, B] MATRIX COMMUTATOR        //
//-----------------------------------------------------------------//
{
    for (int i=0; i<nDIM; i++)
    {
        for (int j=0; j<nDIM; j++)
        {
            commutator[i*nDIM + j] = 0. + I * 0.;
            for (int k=0; k<nDIM; k++)
            {
                commutator[i*nDIM + j] += A[i*nDIM + k]*B[k*nDIM + j] - B[i*nDIM + k]*A[k*nDIM + j];
            }
        }
    }
}


double complex_max_element(cmplx *A, int nDIM)
//----------------------------------------------------//
// 	   RETURNS ELEMENT WITH MAX ABSOLUTE VALUE        //
//----------------------------------------------------//
{
    double max_el = 0.0;
    int i, j = 0;
    for(i=0; i<nDIM; i++)
    {
        for(j=0; j<nDIM; j++)
        {
            if(complex_abs(A[i * nDIM + j]) > max_el)
            {
                max_el = complex_abs(A[i * nDIM + j]);
            }
        }
    }
    return max_el;
}


double integrate_Simpsons(double *f, double Tmin, double Tmax, int Tsteps)
{
    double dt = (Tmax - Tmin) / Tsteps;
    double integral = 0.0;
    int k;

    for(int tau=0; tau<Tsteps; tau++)
    {
        if(tau == 0) k=1;
	    else if(tau == (Tsteps-1)) k=1;
	    else if( tau % 2 == 0) k=2;
        else if( tau % 2 == 1) k=4;

        integral += f[tau] * k;
    }

    integral *= dt/3.;
    return integral;
}


//====================================================================================================================//
//                                                                                                                    //
//                                         FUNCTIONS FOR PROPAGATION STEP                                             //
//                                                                                                                    //
//====================================================================================================================//

void CalculateField(parameters* params)
//----------------------------------------------------//
//   RETURNS THE ENTIRE FIELD AS A FUNCTION OF TIME   //
//----------------------------------------------------//
{
    int i;
    int nDIM = params->nDIM;
    int timeDIM = params->timeDIM;

    double* t = params->time;

    double A_R = params->A_R;
    double width_R = params->width_R;
    double t0_R = params->t0_R;

    double A_EE = params->A_EE;
    double width_EE = params->width_EE;
    double t0_EE = params->t0_EE;

    double w_R = params->w_R;
    double w_v = params->w_v;
    double w_EE = params->w_EE;

    for(i=0; i<timeDIM; i++)
    {
        params->field_out[i] = A_R * exp(-pow(t[i] - t0_R, 2) / (2. * pow(width_R, 2))) * (cos((w_R + w_v) * t[i]) + cos(w_R * t[i]));
    }
}

void CalculateSpectraField(spectra_params* spec_params, int freqINDX)
//-----------------------------------------------------------------//
//   RETURNS THE SPECTRA CALCULATION FIELD AS A FUNCTION OF TIME   //
//-----------------------------------------------------------------//
{
    int i;
    int nDIM = spec_params->nDIM;
    int timeDIM_spectra = spec_params->timeDIM_spectra;

    double* t = spec_params->time_spectra;

    double A_S = spec_params->A_S;
    double width_S = spec_params->width_S;

    for(i=0; i<spec_params->timeDIM_spectra; i++)
    {
        spec_params->field_spectra[i] = A_S * exp(-pow(t[i], 2) / (2. * pow(width_S, 2))) *
        (cos((spec_params->w_R + spec_params->freq_spectra[freqINDX]) * t[i]) + cos(spec_params->w_R * t[i]));
    }
}

void CalculateVibSpectraField(spectra_params* spec_params, int freqINDX)
//----------------------------------------------------------------------//
//   RETURNS THE VIB. SPECTRA CALCULATION FIELD AS A FUNCTION OF TIME   //
//----------------------------------------------------------------------//
{
    int i;
    int nDIM = spec_params->nDIM;
    int timeDIM_spectra = spec_params->timeDIM_spectra;

    double* t = spec_params->time_spectra;

    double A_S = spec_params->A_S;
    double width_S = spec_params->width_S;

    for(i=0; i<spec_params->timeDIM_spectra; i++)
    {
        spec_params->field_spectra[i] = A_S * exp(-pow(t[i], 2) / (2. * pow(width_S, 2))) * (cos(spec_params->freq_spectra[freqINDX] * t[i]));
    }
}

void L_operate(cmplx* Qmat, const cmplx field_ti, molecule* mol, parameters* params)
//----------------------------------------------------//
// 	    RETURNS Q <-- L[Q] AT A PARTICULAR TIME (t)   //
//----------------------------------------------------//
{
    int m, n, k;
    int nDIM = params->nDIM;
    double* gamma_pcd = mol->gamma_pcd;
    double* gamma_decay = mol->gamma_decay;
    cmplx* mu = params->mu;
    double* energies = mol->energies;

    cmplx* Lmat = (cmplx*)calloc(nDIM * nDIM,  sizeof(cmplx));

    for(m = 0; m < nDIM; m++)
        {
        for(n = 0; n < nDIM; n++)
            {
                Lmat[m * nDIM + n] += - I * (energies[m] - energies[n]) * Qmat[m * nDIM + n];
                for(k = 0; k < nDIM; k++)
                {
                    Lmat[m * nDIM + n] -= 0.5 * (gamma_decay[n * nDIM + k] + gamma_decay[m * nDIM + k]) * Qmat[m * nDIM + n];
                    Lmat[m * nDIM + n] += I * field_ti * (mu[m * nDIM + k] * Qmat[k * nDIM + n] - Qmat[m * nDIM + k] * mu[k * nDIM + n]);

                    if (m == n)
                    {
//                        Lmat[m * nDIM + m] += gamma_decay[k * nDIM + m] * Qmat[k * nDIM + k];
                    }
                    else
                    {
                        Lmat[m * nDIM + n] -= gamma_pcd[m * nDIM + n] * Qmat[m * nDIM + n];
                    }


                }

            }

        }

    for(m = 0; m < nDIM; m++)
        {
        for(n = 0; n < nDIM; n++)
            {
                Qmat[m * nDIM + n] = Lmat[m * nDIM + n];
            }
        }
    free(Lmat);

}


void Propagate(molecule* mol, parameters* params)
//----------------------------------------------------------------------//
//    GETTING rho(T)_{k=[3,4]} FROM rho(0) USING PROPAGATE FUNCTION     //
//----------------------------------------------------------------------//
{
    int i, j, k;
    int tau_index, t_index;
    int nDIM = params->nDIM;
    int timeDIM = params->timeDIM;

    cmplx *rho_0 = params->rho_0;
    double *time = params->time;

    double dt = time[1] - time[0];

    cmplx* L_rho_func = (cmplx*)calloc(nDIM * nDIM, sizeof(cmplx));
    copy_mat(rho_0, L_rho_func, nDIM);
    copy_mat(rho_0, mol->rho, nDIM);

    for(t_index=0; t_index<timeDIM; t_index++)
    {
        k=1;
        do
        {
            L_operate(L_rho_func, params->field_out[t_index], mol, params);
            scale_mat(L_rho_func, dt/k, nDIM);
            add_mat(L_rho_func, mol->rho, nDIM);
            k+=1;
        }while(complex_max_element(L_rho_func, nDIM) > 1.0E-8);

        for(i=0; i<nDIM; i++)
        {
            mol->dyn_rho[i * timeDIM + t_index] = mol->rho[i * nDIM + i];
        }

        mol->dyn_rho[4 * timeDIM + t_index] = mol->rho[0 * nDIM + 1];
        mol->dyn_rho[5 * timeDIM + t_index] = mol->rho[0 * nDIM + 2];
        mol->dyn_rho[6 * timeDIM + t_index] = mol->rho[0 * nDIM + 3];
        mol->dyn_rho[7 * timeDIM + t_index] = mol->rho[1 * nDIM + 2];
        mol->dyn_rho[8 * timeDIM + t_index] = mol->rho[1 * nDIM + 3];
        mol->dyn_rho[9 * timeDIM + t_index] = mol->rho[2 * nDIM + 3];

        copy_mat(mol->rho, L_rho_func, nDIM);
    }

    free(L_rho_func);
}


void PropagateSpectra(molecule* mol, spectra_params* spec_params, parameters* params, int freqINDX)
//----------------------------------------------------------------------//
//    GETTING rho(T)_{k=[3,4]} FROM rho(0) USING PROPAGATE FUNCTION     //
//----------------------------------------------------------------------//
{
    int i, j, k;
    int tau_index, t_index;
    int nDIM = spec_params->nDIM;
    int timeDIM_spectra = spec_params->timeDIM_spectra;

    double dt = spec_params->time_spectra[1] - spec_params->time_spectra[0];

    cmplx* L_rho_func = (cmplx*)calloc(nDIM * nDIM, sizeof(cmplx));
    copy_mat(params->rho_0, L_rho_func, nDIM);
    copy_mat(params->rho_0, mol->rho, nDIM);

    for(t_index=0; t_index<timeDIM_spectra; t_index++)
    {
        k=1;
        do
        {
            L_operate(L_rho_func, spec_params->field_spectra[t_index], mol, params);
            scale_mat(L_rho_func, dt/k, nDIM);
            add_mat(L_rho_func, mol->rho, nDIM);
            k+=1;
        }while(complex_max_element(L_rho_func, nDIM) > 1.0E-8);

        copy_mat(mol->rho, L_rho_func, nDIM);
    }

    free(L_rho_func);
//    mol->spectra[freqINDX] = mol->rho[2*nDIM + 2] + mol->rho[3*nDIM + 3];
    mol->spectra[freqINDX] = mol->rho[1*nDIM + 1];
}


void RamanControlFunction(molecule* molA, molecule* molB, parameters* func_params, spectra_params* spec_params)
//------------------------------------------------------------//
//    GETTING rho(T) FROM rho(0) USING PROPAGATE FUNCTION     //
//------------------------------------------------------------//
{
//    mol_system* Ensemble;
//    Ensemble->moleculeA = molA;
//    Ensemble->moleculeB = molB;
//    Ensemble->params = func_params;

//    CalculateField(Ensemble->params);
//    Propagate(molA, func_params);
//    Propagate(molB, func_params);
//    free(Ensemble);

    for(int i=0; i<spec_params->freqDIM; i++)
    {
        CalculateVibSpectraField(spec_params, i);
        PropagateSpectra(molA, spec_params, func_params, i);
        PropagateSpectra(molB, spec_params, func_params, i);
    }
}
